src/cargo/core/compiler/mod.rs

```diff
@@ -393,6 +393,22 @@ fn link_targets<'a, 'cfg>(
         .collect();
     let json_messages = bcx.build_config.json_messages();
 
+    // Find the executable for the current target (if any).
+    let mut executable : Option<PathBuf> = Option::None;
```

Comment:
Wanted to move this inside the job below but ran into problems with `Unit`. Any suggestions of how to resolve these problems are appreciated. Not sure if there are any side effects with the current approach.

Reply:
No, this is exactly how it should look like, and it's kinda cool that Rust detects this at compile time. The `Work` thing which we create below is a long-lived closure, and works are executed concurrently. If in that closure we close over `Context`, than bad things would happen, because several works would try to access Context simultaneously from different threads, and context is not ready for that. 

So Rust enforces that we compute all relevant data before hand, and store just simple stuff like `String`s inside the `Work`.

Reply:
Also, you can use `let mut executable: Option<PathBuf> = None;`, no need to qualify `None`. However, I hope when `bindist` is extracted as a separate function, this could be simplified to `let executable = if produces_executable(unit) { bindist(unit) } else { None }`.

Reply:
@matklad Awesome explaination. Thanks!